.github/workflows/run-qit-extension.yml

```diff
@@ -0,0 +1,154 @@
+name: Run QIT for a given extensions
+
+# **What it does**: Runs a set of QIT tests for a given extension.
+# **Why we have it**: To reuse across other repos, to make a full test of a single extension.
+
+on:
+  workflow_call:
+    inputs:
+      # Basic params.
+      extension:
+        description: Extension to test
+        required: true
+        type: string
+      version:
+        description: |
+          Version to be tested: `latest`, `dev`, or `local`.
+          `latest` will test latest released version,
+          `dev` will make the action look up for an `{extension}.zip` file in the root
+          of the repository under the `gha-dev-build` tag,
+          `local` will look up for an `{extension}.zip` artifact in the current workflow.
+        required: true
+        default: 'latest'
+        type: string
+
+      # Customize which types to run.
+      test-activation:
+        description: 'Should activation be tested?'
+        default: true
+        type: boolean
+      test-security:
+        description: 'Should security be tested?'
+        default: true
+        type: boolean
+      test-phpstan:
+        description: 'Should PHPStan be tested?'
+        default: true
+        type: boolean
+      test-api:
+        description: 'Should API be tested?'
+        default: true
+        type: boolean
+      test-e2e:
+        description: 'Should E2E be tested?'
+        default: true
+        type: boolean
+
+      # Advanced customization.
+      options:
+        description: 'Additional options for `qit` command, like `--optional_features=hpos`'
+        type: string
+
+    outputs:
+        statuses:
+            description: "Statuses of all tests. Array of integers."
+            value: ${{ jobs.qit-tests.outputs.statuses }}
+
+
+jobs:
+    qit-tests:
+        name: Run QIT Tests
+        runs-on: ubuntu-20.04
+        env:
+            QIT_DISABLE_ONBOARDING: yes
+            dev_build: ${{ inputs.version != 'latest' && format('{0}.zip', inputs.extension) || '' }}
+        outputs:
+            statuses: ${{ toJSON( steps.*.outputs.status ) }}
+        steps:
+
+            - name: Download `gha-dev-build`
+              if: ${{ inputs.version == 'dev' }}
+              uses: robinraju/release-downloader@v1.8
+              with:
+                repository: "woocommerce/${{ inputs.extension }}"
+                tag: 'gha-dev-build'
+                fileName: ${{ env.dev_build }}
+                token: ${{ secrets.BOT_GH_TOKEN }}
+
+            - name: Download artifact
+              if: ${{ inputs.version == 'local' }}
+              uses: actions/download-artifact@v3
+              with:
+                name: ${{ env.dev_build }}
```

Comment:
Yes, this is the way to use it.


> A concern is the workflow files in `.github/workflows` directory of this repo won't be included in the actions-v1 releases. As a result, it would become difficult to ensure version compatibility between them.

I completely agree with that concern. That's why in the OP I mentioned:
> I was thinking of pushing it to `github-actions` folder. So, we will have it under common release flow and location, with more convenient version tags. But I'm not sure, as this is "workflow" not "action". Maybe we can store it in a location like `.github/workflows/reusable` or `packages/workflows` and preserve that folder in the github-actions release?

I would appreciate your opinion here. I'd say having separate release flows for workflows and separate for actions is too much. I would choose one of those approaches:

- Consider our _[github] <code><strong>actions</strong></code>_ tag actually mean _"github **actions & workflows**"_, then
   1. consider workflows and atomic actions completely equal, and make consumers distinguish what is what:
   ```yml
   jobs:
     build:
       runs-on: ubuntu-latest
       steps:
        - name: Reuse action
          uses: woocommerce/grow/prepare-php@actions-v1 # Action - should be used in `jobs.*.steps.uses`
     runQIT:
       name: Reuse workflow
       uses: woocommerce/grow/run-qit-extension.yml@actions-v1 # Workflow - should be used in `jobs.*.uses`
       needs: build
   ```
   2. put workflows and actions into distinct folders (so maintainers and consumers can easily distinguish what is what)
   ```yml
   jobs:
     build:
       runs-on: ubuntu-latest
       steps:
        - name: Reuse action
          uses: woocommerce/grow/action/prepare-php@actions-v1
     runQIT:
       name: Reuse workflow
       uses: woocommerce/grow/workflow/run-qit-extension.yml@actions-v1
       needs: build
   ```

> To run QIT tests with a local build, I would suggest leaving the processing of preparing a local extension zip file on the action user side. For example, GLA could either build a zip file within its workflow or run the actions/download-artifact step on its own, and this action only needs the local file path as input.

The problem here is that. This is a workflow (`jobs.*.uses`), not an atomic action (`jobs.*.steps.uses`). This workflow runs as a separate job, from any other workflow or action. So the only way to share a file (I'm aware of) between this workflow and another (GLA's, or other extension's) workflow is to use artifacts. Then this workflow needs to download them itself. Otherwise, how would GLA's job push a file to this job?

Reply:
The version compatibility concern in https://github.com/woocommerce/grow/pull/81#issuecomment-1743623849 is the same as this thread.

> > The current way of transferring a local extension zip might not be easy to handle in the future in terms of version compatibility. Wonder if it would be better to leave the processing to the action user side.
> 
> Could you elaborate on that?

Sorry, I omitted a large portion of explanation. The idea is to transform the `run-qit-extension.yml` workflow into a composite action and it involves a degree of change that is not quite easily stated in words. I created a demo in the repo and one in GLA to illustrate how it works.

### [Changes in this repo](https://github.com/woocommerce/grow/compare/add/qit-workflows...add/qit-workflows-alternative)

- Based on the `run-qit-extension` workflow, create a composite `run-qit-extension` action.
- Remove the "Download artifact" step.
- Add `qit-partner-user` and `qit-partner-secret` inputs to forward the original secret variables.
- Remove all `timeout-minutes` as they are not supported in the composite action. Probably it would be better to let the action user specify the timeout.
- I didn't transform the `outputs.statuses` as it would need more time or the composite action might not support the same syntax.
   - https://github.com/woocommerce/grow/blob/a8f23bae23cbe0e896189ec561099897622e4ac1/.github/workflows/run-qit-extension.yml#L52-L55
   - https://github.com/woocommerce/grow/blob/a8f23bae23cbe0e896189ec561099897622e4ac1/.github/workflows/run-qit-extension.yml#L65-L66

### [Changes in GLA](https://github.com/woocommerce/google-listings-and-ads/compare/dev/qit-workflow...dev/qit-workflow-alternative)

- Add the "Download artifact" step.
- Use the `run-qit-extension` action instead.

---

The test run: https://github.com/woocommerce/google-listings-and-ads/actions/runs/6391261008

I didn't continue to make further changes. If adopt this method, the `.github/workflows/run-qit-all.yml` workflow should be able to use `run-qit-extension` action as well, and the `.github/workflows/run-qit-extension.yml` workflow would be no longer in needed.

In this way, the action user sides won't need to use this repo's workflow directly, and this repo doesn't need to add them to any release and versioning flow.

Another advantage is that the action user side wouldn't have to upload artifacts if their action builds a local zip file and uses this action directly. For example:

```yaml
# ... omitted

jobs:
  qit-tests:
    name: Run QIT Tests
    runs-on: ubuntu-20.04
    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      # ... omitted

      - name: Build production bundle
        run: |
          npm run build

      - name: Delegate QIT Tests
        uses: woocommerce/grow/run-qit-extension@actions-v1
        with:
          qit-partner-user: ${{ secrets.QIT_PARTNER_USER }}
          qit-partner-secret: ${{ secrets.QIT_PARTNER_SECRET }}
          version: 'local'
          extension: 'google-listings-and-ads'
          test-security: 'true'
```

Reply:
Thank you for the detailed explanations and the new branch! 
I can see it works, and I appreciate how it addresses the following:

> this repo doesn't need to add them [workflows] to any release and versioning flow.

However, I'm not sure if I see all the benefits of 
>  the action user sides won't need to use this repo's workflow directly

- I get that accessing the workflow directly is not great, and I'm happy to improve it. But I think we can to it also otherwise
- Using action or workflow from code perspective is not much longer


The biggest difference I see is granularity. Having `qit` as an atomic action, users could compose it within their workflows.
But I think it also has the downsides:
- the log is less collapsible and divided.:
as a workflow: 
![image](https://github.com/woocommerce/grow/assets/17435/4589d116-4646-46f4-b41d-f17a557ab270)
as an action
![image](https://github.com/woocommerce/grow/assets/17435/5fa742b5-eb21-4c2d-83e1-6ef18ab0eb1f)

- I was actually thinking of unifying the workflow across our repos so we will not have to implement a workflow and compose actions in every repo but just use the shared entire workflow.


Especially if it comes to the build, I think it is even desirable to have it as a separate workflow that uploads an artifact. So, other workflows and actions running for the same commit could reuse that work, reducing the overall cost and time.
See https://github.com/woocommerce/google-listings-and-ads/pull/2114#discussion_r1341025517

Reply:
> * I was actually thinking of unifying the workflow across our repos so we will not have to implement a workflow and compose actions in every repo but just use the shared entire workflow.

From the perspective of this repo, I would suppose its granularity prefers a [custom action](https://docs.github.com/en/actions/creating-actions/about-custom-actions) over a [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow) for flexibility in composition and configuration and fewer limitations.

If an automation requirement is highly repetitive or identical across several repos, then a reusable workflow, which might still composite some custom actions within this repo, would be appropriate.

> Especially if it comes to the build, I think it is even desirable to have it as a separate workflow that uploads an artifact. So, other workflows and actions running for the same commit could reuse that work, reducing the overall cost and time. [...]

From a practicable point of view, I find it rather difficult to imagine whether it would be easier to do so.

Take GLA as an example, given it has a reusable workflow for building a zip bundle, and each automation of "Bundle watch", "Publish dev build", "E2E test", and "QIT" need a build file as input.

Unless all these automations are in the same workflow, each automation will be grouped into standalone workflows with the same execution conditions, and then build a zip on its own. Because artifact access through `actions/upload-artifact` and `actions/download-artifact` is limited within the same workflow run. In this case, the same commit doesn't use the "same" work result (zip bundle), only the extracted build workflow is reused.

However, having all of them in the same workflow run seems to make that workflow configuration very complex.

Reply:
My main concern is still how the release versions of these reusable workflows in the `.github/workflows` directory will proceed if go with the current solution.

The unintuitive understanding of the artifact upload/download split between the use-end repo and this repo is also a concern, but if this complexity trade-off is necessary, then it won't be a blocker.

> I'd say having separate release flows for workflows and separate for actions is too much.

Agree with this.

> I would choose one of those approaches: [...]

GitHub docs say the syntax of [`jobs.<job_id>.uses`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_iduses) is `{owner}/{repo}/.github/workflows/{filename}@{ref}` and doesn't clarify if the `.github/workflows/` part can be changed.

Either yes or no, I lean toward leaving reusable workflows under the original directory and keeping custom actions the same as it's:

```yml
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
     - name: Reuse action
       uses: woocommerce/grow/prepare-php@actions-v1

  runQIT:
    name: Reuse workflow
    uses: woocommerce/grow/.github/workflows/run-qit-extension.yml@actions-v1
    needs: build
```

The main reason is that each of them conforms to the syntax described in GitHub docs respectively.
- [`jobs.<job_id>.steps[*].uses`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsuses)
- [`jobs.<job_id>.uses`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_iduses)

Reply:
I merged your proposed changes and used action as you suggested.
We can get back to shared workflows if needed in some following PRs, exploring possibilities and flow optimizations there.